### DISCUSSION TITLE
Storage of encrypted passwords in homeassistant configuration file

### DISCUSSION THREAD

**Discussion by cnico** [home-assistant/architecture#556] [General]
Hi,

As discussed in this pull request : https://github.com/home-assistant/core/pull/46582, I fall on the fact that homeassistant stores the password in its configuration in clear text.
It was quite surprising to me and seems to be avoided so I did a small encryption of it before and after reading the configuration.

@iMicknl suggested to have the discussion here about this behavior and if this solution should not be directly a general purpose function in the core api.

What do you think of it ?

Regards,

**Comment by frenck** [comment]
Please provide a full proposal and details in this discussion (and thus not referring to the implementation for details).

**Reply by iMicknl** [reply]
Related to https://community.home-assistant.io/t/managing-secrets/269485 as well.

**Reply by cnico** [reply]
I read the conversation and discovered that the need of password not being stored in plain text in HA config files has many options discussed as of solutions that are more or less robusts.
I do not think I am enough in expertise to have pertinent expression on these solutions as I do not know homeassistant's core internals and architecture details.

My only possible contribution is to tell my opinion : homeassistant should store passwords in encrypted string.
Thinking of it globaly, it should be transparent to integrations developpers with a mechanism that automatically encrypt config values behind the key CONF_PASSWORD. It currently is the case in frontend when the user enters it.

This said, I also think that, in agile design point of view, the security robustness could be implemented with iterations of basic to more complex solutions with homeassistant releases.
A first release could be : 
- encrypt / decrypt passwords writing/reading configurations with a simple pycryptodome method.
- the seed (or master key) could be stored in a hidden file accessible to only homeassistant's root user and not accessible to integrations.
- a batch of existing passwords could be implemented to encrypts all already existing ones

A second release could be : 
- usage of a vault as some sort of mandatory addon that would store the effective passwords in a more secured manner
- the seed or master key would be still a challenge to store automatically has in first release. I would opt for the same solution for this point.

A third release could be : 
- allow the user to choose the storage form of its master key for passwords : for example, asking him at installation and never stoing it in a file if he wants to secure its home assistant installation. The password could be entered at startup manually or through some device like a yubico.


What do you think @frenck , @iMicknl ?

For my flipr integration, I decided to remove the code I did in the idea first release explained here considering it will some day be done directly in the core by far more experts than me.

**Reply by spacegaier** [reply]
As Jc2k already mentioned below, your release 1 proposal does not really add real security (see first answer from Franck in the above linked thread). It might suggest to the user that something is more secure than it actually is ("hey I see it's encrypted"), since there is still a decrypt key lying next to the "encrypted" passwords.

**Reply by RichieFrame** [reply]
My suggestion:

Do not stagger releases, go full secure from the start

HA would have a master encryption 256-bit key, and for each integration entry requiring encryption, generates a nonce.
HA would also have a master config authentication key.
The integration's config flow would need to flag which fields in the integration are to be encrypted.
By default, all password fields in all entries could also be encrypted.
The config flow generates a secure 128-bit key for the integration entry, and for each field requiring encryption, generates a nonce.
When the field needs to be stored, it is encrypted with the entry key and the per-field nonce using AES-128 in GCM mode.
The entry key itself is encrypted using the master key and the per-entry nonce using AES-256 in GCM mode.

example of a config entry in its current state and a fake encrypted version 

```
            {
                "entry_id": "561e76701746438d94619d60e617a85e",
                "version": 1,
                "domain": "onvif",
                "title": "LOREX 181",
                "data": {
                    "host": "192.168.0.181",
                    "name": "LOREX",
                    "password": "password",
                    "port": 80,
                    "username": "admin",
                    "snapshot_auth": "digest"
                },
                ...
             }
```
```
            {
                "entry_id": "561e76701746438d94619d60e617a85e",
                "encryption_key": {
                    "version": 1,
                    "nonce": "cb584e44c43ed6bd0bc2d9c7",
                    "value": "PG4LipwVIkqCKLmpjKFTHQ",
                    "tag": "5e9b60f69165f32f8930843c"
                },
                "version": 1,
                "domain": "onvif",
                "title": "LOREX 181",
                "data": {
                    "host": "192.168.0.181",
                    "name": "LOREX",
                    "password": {
                        "encrypted": true,
                        "encrypted_data": {
                            "version: 1,
                            "nonce": "b4d446e2fc56e5ac33639e50",
                            "value": "zUJATVKtVcz6mspK3IKKpYAK2dOFoGcfvL9yQRgyBhk",
                            "tag": "7d86d871880a8d7145f9cd88"
                        },
                    },
                    "port": 80,
                    "username": "admin",
                    "snapshot_auth": "digest"
                },
                ...
             }
```

Reading the password field would see that it has an encryption flag, then it would read the entry key, decrypt and verify with the master key, then use the result to decrypt and verify the field value. The field data values in the example are Base64url encoded, the nonce and tag are hexadecimal. Adding a version allows easy changes to the encryption engine in the future. Current practices suggest a 128-bit key is secure for the next decade, and the parent key should be double the size of a child key if using multilayer encryption.

The specific hierarchy above provides confidentiality and authenticity of the data, exposure of field plaintext does not compromise the entry key, and exposure of the unencrypted entry key does not compromise the master key. The GCM algorithm should accept the domain and entry_id of the entry as well, to prevent a simple copy of the encryption related data to another entry, even of the same type in the same domain.

Above that, any config entry that has been given a key should have the entire entry data digested with HMAC using the master config authentication code, and the result stored in a secure file. This would prevent someone changing something like the ip address or hostname, in order to tunnel the decrypted field data in any form to a specific attacker controlled device on the network or internet. Even a 1-bit change to the configuration would result in an authenticity failure.

If the config flow class takes care of all the field level stuff, integrations using config flow should not require any modification to their code... I assume. The encryption engine would then be added as an intermediate step before it is written to the config. Instead of the password as text, it writes the JSON output of the entry encryption engine, which includes the flag and the data. The master encryption engine outputs the same data in the same format less the flag, only the key size is different. Both engines would easily be created from the same class.

As far as I can tell, pycryptodome will perform all the required operations except for Base64, which is built into python. Using those, very minimal code would be required to create the class, the storage format above is easy to read and highly tamper resistant, requiring the secure storage of only 2 master keys.

There would need to be a discussion regarding how this secures against various threat models, if the encryption engine code would be simple enough for eyes to see any potential shortfalls, how much work is needed to intercept the data to/from the integration and its config flow, etc.

**Reply by RichieFrame** [reply]
Here are my new examples of encrypted structure, I made a few changes in the name of simplicity of mapping, so there is less checking the structure, and it is easy to map between and encrypted and plaintext state.

This example was generated using the proof of concept code, which is now fully functional for decryption and encryption of config entries. Only 2 functions need to be exposed to the ConfigEntries class to use it, the encryption function which takes the config entry and a list of fields to encrypt, and the decrypt function.

Before encryption:
```
{
    "entry_id": "561e76701746438d94619d60e617a85e",
    "version": 1,
    "domain": "onvif",
    "title": "LOREX 181",
    "data": {
        "host": "192.168.0.181",
        "name": "LOREX",
        "password": "password",
        "port": 80,
        "username": "admin",
        "snapshot_auth": "digest"
    }
}
```

After encryption:
```
{
    "entry_id": "561e76701746438d94619d60e617a85e",
    "version": 1,
    "domain": "onvif",
    "title": "LOREX 181",
    "data": {
        "host": "192.168.0.181",
        "name": "LOREX",
        "port": 80,
        "snapshot_auth": "digest"
    },
    "config_encryption": {
        "enabled": true,
        "key": {
            "version": 1,
            "nonce": "918f89e544b2fa38b06bc1f1",
            "value": "tuBasOfxhu4PbhFSK5PDdA==",
            "tag": "ea4084b164d8f1348f5da023"
        }
    },
    "data_encrypted": {
        "password": {
            "version": 1,
            "nonce": "e2a4de2cba936ab483e410d3",
            "value": "ljGoEOKQzpI=",
            "tag": "acb15dd4c2d98c76dccc78a9"
        },
        "username": {
            "version": 1,
            "nonce": "3bc683b5f30a711f9364ddbf",
            "value": "MCv4UUI=",
            "tag": "b0ba21cbe7e326b73dff4fb2"
        }
    }
}
```

It is obviously a little different from the prior example in structure, this prevents any sort of potential issue reading out of the data key, since everything in data_encrypted is now guaranteed encrypted. Some key names are also changed. 

I was also able to keep the same functional encryption code for both the data keys as well as the config_encryption key, plus 2 functions that encrypt and decrypt the config_encryption key that would have access to the master crypto key.

Not included in the proof of concept but of possible use is a padding function to hide the true length of the values prior to encryption, since that can be considered sensitive information. It would pad to a length of increments of 16, plus overhead.

The proof of concept has no provision for storage or management of the master crypto key, since I am unfamiliar with how HA should handle that. It also has no authentication for the entire config entry, but the entry_id and domain are authenticated in addition to the encrypted fields. That means direct modification of the plaintext or encrypted data keys will not raise an error.

Once the encrypted entry is read from disk into a python data structure, it needs to be decrypted:
```
{
    "entry_id": "561e76701746438d94619d60e617a85e",
    "version": 1,
    "domain": "onvif",
    "title": "LOREX 181",
    "data": {
        "host": "192.168.0.181",
        "name": "LOREX",
        "port": 80,
        "snapshot_auth": "digest",
        "password": "password",
        "username": "admin"
    },
    "config_encryption": {
        "enabled": true,
        "key": {
            "version": 1,
            "nonce": "918f89e544b2fa38b06bc1f1",
            "value": "tuBasOfxhu4PbhFSK5PDdA==",
            "tag": "ea4084b164d8f1348f5da023"
        }
    }
}
```

The data_encrypted keys are decrypted, then plaintext inserted into data. Then data_encrypted is removed, and the function returns the decrypted entry as above. I did not remove the config_encryption key, since it can provide information to the integration that HA is performing crypto operations on it. It could also be useful to retain the data_encrypted key, either unmodified or a skeleton like so:

```
    "data_encrypted": {
        "password": {
        },
        "username": {
        }
    }
```

I am going to clean up my code, add some test examples, and post to github so everyone can play with it, see if any changes should be made, how to integrate it, and decide if it is something that is useful and should be added as a feature.

**Reply by frenck** [reply]
The problem remains the same, it still can be accessed directly a CVE can be reported against this structure instantly.

There is absolutely not point in doing it, as everything still can access it, and can decrypt it.

> The proof of concept has no provision for storage or management of the master crypto key, since I am unfamiliar with how HA should handle that. 

Well, that...

**Reply by RichieFrame** [reply]
The example above has no master key, there is **absolutely no way to decrypt the data without it**... And that master key is the only one that needs to remain secured, presumably the same way other confidential authentication information is stored. That is the point of the structure and its hierarchy. You need the master key to get the entry key, and you need the entry key to get the encrypted data. Therefore you need the master key to get the encrypted data.

Hell, you could even use a HSM to do all of the encryption if you wanted so no key material ever makes it to the OS, and store said master key on the HSM... though the cheapest one I found which supports modern encryption algos is about $700 USD.

either way, here is some code if anyone wants to look or play:

https://gist.github.com/RichieFrame/b6fddc6f8e01b883f042b5130f954e13

**Reply by frenck** [reply]
The master key needs to be with the system... else you can't start. HSM keys are not really an option...

As long as Home Assistant can access the master key, it still has no use, as custom/arbritary code runs inside Home Assistant.

**Reply by RichieFrame** [reply]
being with the system does not mean that HA cannot access it in a way that prevents custom components and third party modules from accessing it. How is this for a suggestion:

The master key is sent as a command line argument to python, the internal command line is not an immutable list, once it is extracted by main() using get_arguments(), the list is cleared (since it is no longer required) and nothing beyond can access the list. Now the key just needs to be put in memory until the crypto class can be loaded, and moved there, where it stays. My understanding of python is not elite, but I am pretty sure something like that is possible.

Obviously that is an extreme oversimplification, it could be that the argument is a path to a 400 permission root owned temporary file that was populated with the key, and it reads it out of the file and then erases it, or some other form of IPC. Either way, the storage and securing of the key are then no longer part of the HA process, and if HA is run as a non-root user, there is no way it could read the file even with insecure directory traversal.... right?

obviously a root access user is going to be able to get to that key file anyway, but securing the filesystem from an attack outside of the operating context should not be part of this threat model.

**Reply by frenck** [reply]
> The master key is sent as a command line argument to python

That is already a no go. That doesn't work.

**Reply by RichieFrame** [reply]
Is it a technical reason why that wont work, architectural? With minimal modification to 3 of the core files I was able to get the logger to print out a string passed using a --master-secret argument. Of course that shows up in the OS command line context so passing a secret directly is ill advised, but the path to a secret outside of the config dir with the correct permissions?

**Reply by Jc2k** [reply]
Architecture issues aren't purely technical. The maintainers are trying to balance value to users (and to how many users). And also to try and keep technical debt under control. HASS is a massive code base, and their team is small. They also need to think about user experience. So there are many reasons why even perfect code might be unacceptable.

You have put a lot of thought and care into your scheme, but I think the hard bit is proper handling of the master secret. If we can't manage that in line with the requirements of your threat model then it will undermine all the thought and care elsewhere in your model.

HASS is something that runs on $30 computers under the stairs. Most users don't want to buy expesnive HSMs, we don't have secure boot, we don't generally have TPM's. So I think the assumption is that it will be hard to do something like this well, and for users where this matter full disk encryption might be a better choice.

I don't think we have a shared agreed threat model, so here is what I think the thinks we could stop are, and the things we can't stop. Not a full threat model, but hopefully something we can agree on. From what I can tell, encryption here only helps protect the secrets at rest. There are a couple of "threats" I can think of:

* Physical theft of the device. The secret key does not exist on disk, only in non-swappable memory.
* A CVE that allows an attacker to read static files on the local disk, but nothing else. Again the secret key does not exist on disk.

It can't help with:

* A CVE that allows code execution, as the secrets will be decrypted in memory.

And there are some caveats:

* I haven't looked in a long time but i seem to remember it's hard to `mlock` in python. At the time I think the options were using a custom C-extension, using encrypted swap or disabling swap all together. 

So to not ruin the benefits your encryption scheme afford we need to take the encryption key at runtime. That means HA can't properly start until someone has armed it with a secret key. We could maybe do this in browser, but not everyone uses TLS.

I think if we did this we would want everyone to benefit from it, rather than adding a configuration option that only 10 users tried out. Otherwise is the maintenance worth it. So this needs to consider what the user experience is. For example, we don't want our beginner friendly images requiring you to SSH in to run `hass --master-secret mysupersekritpassphrase`.

**Reply by RichieFrame** [reply]
I agree with that information. I assume that any technical solution should not require specialized hardware, that was only mentioned in passing since it is a technically possible solution to an extreme threat model. Additionally, a solution should work on hassos, docker, and supervised, and optimally core as well, though core might be difficult. Also some things like tempfs require a linux host leaving windows users out, but that may be a small minority.

Possible threat models include:
In person attacks:
--Hardware theft
--Drive cloning
--Evil maid (system remains online)
Attacks over the network, either local or internet
Exploits allowing reading from .storage that a malicious extension could access
Exploits allowing reading from the root file system that a malicious extension could access

Due to the nature of HA, a malicious extension taking place of an integration to read its config should not be in the model, since working around that would naturally require digital signatures, and a complex infrastructure to support them, in addition to a web of trust. Think how firefox started requiring signed extensions a few years back.

Also it should be able to be accessible after an unexpected reboot of the machine, restart of a container, etc

Depending on the threat level and model, a secret location can be:
in a file inside the config directory
in a file inside the .storage directory
in a file somewhere else on the host operating system
available through a running application on the host system
available through a running containerized service

If you look at how something like wordpress or apache stores a secret, or a server stores a tls private certificate file, they are typically just there on disk, but with correct permissions from an operating system context that an unprivileged user cannot access them. If that system has FDE controlled by a TPM, a power cycle has no impact, server comes up, program runs, reads the file, and the system is protected from offline drive cloning.

I would suggest the threat model used here does NOT take hardware theft and drive cloning into account, if someone wants to FDE with a TPM, that level of protection is outside the scope of config encryption and should be up to the user. We should focus on directory/file system traversal.

For containerized/hassos situations on linux, you could have the container mounted with a tempfs that contains a copy of the key, it is loaded into HA during the bootstrap process, then it gets deleted or unmounted. Since it was a copy, the real key still exists outside the container for when it needs to run again. Now it is no longer accessible anywhere in the running process' file system.

My suggestion would be to keep a master secret outside of the config folder, but to enforce storage of a password protected copy there, so that a backup recovery to a new system can provide the secret. That way exposure of the entirety of /config will not expose any authentication data for integrations. 

My next level of test would be on the docker side, get the secret into the container through a tempfs, and see if i can delete it through the HA runner. I could also use a docker secret, but then I need to convert the instance to a swarm service, and I am not ready to do that just to test at this point.

**Reply by frenck** [reply]
I'm out of this, as I think you are not hearing it 👋

**Reply by RichieFrame** [reply]
Jc2k brought up good points that warrant discussion, but what I heard from you is something along the lines of "that wont work" or "if HA can access the key it is insecure" with no explanation, even in short detail, of why. I think it boils down to this sentence: 

> As long as Home Assistant can access the master key, it still has no use, as custom/arbritary code runs inside Home Assistant.

I do not think that is the case, arbitrary code only runs up to a certain point, and not everything in the program should be available to everything else, are you telling me that ANY sort of confidential data can be accessed by any malicious custom integration?

**Reply by frenck** [reply]
Its Python.

**Reply by Jc2k** [reply]
TBH, I think the same as @frenck, I'm just using more words because I want you to understand why we think its a lot of work for very little return, and that using host level security features might be a better use of time.

I think your example about TLS private keys is interesting because it demonstrates the bootstrapping issue we face here. Apache out of the box uses user seperation as a security boundary to protect the key material. But whilst you can have a password protected key file, I would say it is very uncommon. I don't believe they have solved having the passphrase on the disk with the server, and require you to type it in at boot. And presumably after patching. Which is why it is uncommon, it can hurt your service's availability whilst providing little that user seperation don't.


>>  As long as Home Assistant can access the master key, it still has no use, as custom/arbritary code runs inside Home Assistant.

> I do not think that is the case, arbitrary code only runs up to a certain point, and not everything in the program should be available to everything else, are you telling me that ANY sort of confidential data can be accessed by any malicious custom integration?

So in order to make a strong and secure web sandbox browsers have to use process seperation with strict and minimal IPC's. We don't have anything like that, and there isn't even such a thing as a strong python in-process sandbox, and if there was we don't use it. A custom integration can totally access anything it wants or even monkeypatch running code that could subvert the encryption process, even by accident.

At the end of the day its python, it can't even pin the passphrase in memory with `mlock` without a custom extension.

**Reply by Jc2k** [reply]
(Remember that even if we call an integration and don't pass it any of our globals, it can use the `gc` module to find every object that is currently alive, so we are pretty much screwed against adversarial code).

**Reply by RichieFrame** [reply]
I understand that python is an interpreted language, I was under the impression that interpreters and runtime environments generally have some kind of "production mode" that turns off debug/dev features, sometimes as the default setting, and it wont allow arbitrary memory dumps from inside the application... is that not the case with python?

After some looking 3.8 added runtime auditing, which could be useful at say, preventing a module from importing gc or running exec. At what point does this turn into a discussion about general security hardening?

**Reply by frenck** [reply]
> At what point does this turn into a discussion about general security hardening?

? Its Python... You are in the wrong place for that discussion.

**Reply by Jc2k** [reply]
Python expects that you trust the code you run, and it is notoriously difficult to sandbox. Projects that do it successfully have to make a very minimal runtime that is hard to recognise as python any more, and even then they get hit with CVE's. Consider:

```
().__class__.__bases__[0].__subclasses__()
```

As long as you can instance an empty tuple, you can access every class that is currently defined in the running HA process. That is a design feature, as python is an introspectable language. You can of course replace `()` with `""`, `[]` or `{}`, or any legitimate object that an integration has access to.

The audit approach looks on the surface like it might actually work. That isn't the case. You can turn on auditing and see something called `open`. But you can't see which integration triggered an open, just that *something* did. If you tried to be clever and figured that a stack trace might let you work out if an integration is involved, unfortunately `sys._getframe` is an audit event and now you are stuck in a loop.

(You could make it so that the secret file was only read once using the audit module, but as discussed custom integrations would be able to fish it out of memory).

There are only a handful of audit events and using one of the tricks like the one above doesn't actually trigger any audit events. Even accessing a juicy module out of `sys.__module__` directly does not generate an audit event. 

You also need to review all your dependencies (and their dependencies dependencies dependencies) because one of them might inadvertently wrap a C API that can be used to do io and completely bypass any auditing hooks. Which they might do for legitimate performance reasons - C code can release the GIL so that threaded code can actually be performant.

Of course as soon as a custom integration doesn't work with auditing turned on (because its actually a bad integration for example) people will start advising each other to turn auditing off because it breaks everything.

Regardless of language, if you genuinely wanted to trust untrusted integrations I think the only approach i'd consider would be what browsers do. Each integration in its own process, perhaps in a https://wiki.archlinux.org/title/Bubblewrap sandbox. You can achieve this to some extent without changing core by making your integration a homebridge plugin and interfacing with homekit_controller to control it from HA. Or by implementing the ESPhome websocket API and pretending to be an ESPhome accessory. That sort of approach starts with a really minimal sandbox and gradually adds needed holes. Which is always preferable to trying to fill the existing holes back in.

Anyway, this is off topic. I don't think I have anything else to bring to this topic beyond what i've said. To reiterate, you've identified some things you can protect against. Please try to think about the user experience. I'm sure you don't plan to have a user log onto their HASS machine via ssh, create a tmpfs, write the secret in there with vim and then start HASS with the tmpfs bind mounted. So how will that work, and how will unattended booting work. Without a really comprehensive UX plan I don't think you will get traction.

**Reply by frenck** [reply]
Agree. Encrypting/Decrypting is not the issue, like many methods available to do that, which is more a detail. But if added / think about doing anything:

- It should be user-friendly to use for my dad even... Regardless of how one runs Home Assistant.
- The encryption secret should not be written next to it (e.g., "the-post-it-on-the-monitor), and should not need user interaction to apply on each start or run.
- It should not provide a false sense of security.
- I consider runtime protection impossible with Python, so maybe that needs to be out of scope, but that would be the main reason to add it I guess. That said, if Pyton decrypts it, so in the end... I think this is an endless loop with chicken and egg.

And... we had this, we had support for keychains and vaults, these have been removed, as they prove to be impractical and actually... nobody used them (or cared about them as no-one complained even mentioned anything about the removal).

**Comment by Jc2k** [comment]
The problem with this sort of change is making sure its not security through obscurity. And that can be surprisingly hard to do as i'm sure you are aware! To that end, I think the proposal would benefit from a threat model to review proposals against.

E.g. if your goal is to protect the secrets from physical theft, the approach described in the release 1 would obviously not help. I could simply mount your drive/disk on another system as root and get the secret. If your goal is to protect against someone using a code execution bug against home assistant, it would not help as the attacker can easily dump the memory of the process.

So I think to properly discuss this we need to be reviewing the proposed solution in the context of which threats are in scope and which threats are out of scope.

**Reply by cnico** [reply]
Yes, you are right but I prefer as you call it obscurity to clear text because it will discourage most of the people to try to find the real text whereas in other case they have it right under their eyes.

If we compare it to real world, people buy camera or alarms in their house thinking it is securing their house but a good burglar will have wifi or GSM jammer so that alarm is indeed useless for such attackers.
My point of view is the same here : I do not try to prevent a federal agency to try to still my secrets (if they attack me, they will have more to find out of homeassistant) but a basic user that will share its configuration on a forum when asking for help or a bad integration or custom component that will try to steel passwords of other integrations. 
People store their passwords of cloud providers in homeassistant and I think a minimal obscurity is a good step.

**Reply by Jc2k** [reply]
A threat model is exactly about making sure you don't waste energy protecting against the wrong thing (like a nation state).

But I think if the feature was advertised as secret encryption users would think it was secure against many different things. Most likely against physical theft. So it does need to be clear what it's for and not for.

So far this encryption is just to protect against self owns (where you paste something on the forums looking for help). Do you have any other use cases for it?

**Reply by cnico** [reply]
Actually not, I think it could also help malicious integrations (including via HACS) to steel password of other but there is perhaps some sort of design sandboxing I do not know between integrations. I did not try to hack in this domain.

The main point is just the principle to store password in plain text which I consider a globaly bad practice : this is a question of point of view.

But if the core team decide to do nothing on this point, it is not a big problem for me : I will continue to use homeassistant :-).

**Reply by balloob** [reply]
If you can run code inside Home Assistant (a malicious integration), you're lost already.

**Reply by thecode** [reply]
As others stated here I don't think this should be considered a solution about one integration steeling data from others, a malicious integration could even replace the original one.
However it might be a good as protecting saving passwords in clear text for example.
The mechanism above using `encrypted_data` looks too complicated for me. What I think should be implemented is a simple helper for decrypting/encrypting data for using in integrations, so that if an integration wants to store encrypted data it will just use the helper and upon reading the data it will use the helper again to decrypt.
Your example above will look like this:
```json
{
    "entry_id": "561e76701746438d94619d60e617a85e",
    "version": 1,
    "domain": "onvif",
    "title": "LOREX 181",
    "data": {
        "host": "192.168.0.181",
        "name": "LOREX",
        "password": "zUJATVKtVcz6mspK3IKKpYAK2dOFoGcfvL9yQRgyBhk",
        "port": 80,
        "username": "admin",
        "snapshot_auth": "digest"
 }
```
upon reading the password field the intergration can call helper from `homeassistant.helpers` for example:
`decrypt(entry.data[CONF_PASSWORD])`
And for storing:
`entry.data[CONF_PASSWORD] = encrypt("password")`

**Reply by frenck** [reply]
There is no win in such a method, as anything can run that.
It's like sticking your password on your computer screen.

**Reply by RichieFrame** [reply]
looking complex and being complex are not the same thing, especially when it comes to crypto. There are minimal requirements to doing things correctly. All of those things in my sample still need to be there, you are just mashing them all together in a single string of text for "simplicity". Additionally consider the following:

in 2021.10 config_crypto is added, and a few integrations support it.
in 2021.11, a user's integration adds support, and encrypts the password.
but... something goes wrong with some other thing, and they roll back to 2021.10
now the integration is broken, even though that version of HA has config_crypto support

That happened because it was controlled by the integration, instead of HA. Adding it to HA in the config_entry code and having full decrypt support for 2 or 3 releases before allowing integrations to flag fields should prevent any end-user issues that break the integration configuration. Though as I suggested, there should be no changes to an integration required to support a future change where specific fields names are encrypted by default, "password", "secret", "aws_secret_access_key", etc.

I am writing a proof of concept with a few structural changes right now, it is going well, at some point someone with detailed knowledge of how the config entry IO is done would need to step in and show how this code could interface, though I took a look at it briefly and might be able to figure it out at some point to do a live test.
